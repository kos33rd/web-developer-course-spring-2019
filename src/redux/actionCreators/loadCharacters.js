import axios from "axios"
import * as actionCreators from "./index"


export const loadCharactersActionCreator = (values = {}) => (dispatch) => {
    const { query, aliveOnly } = values

    const queryParams = []

    if (query) {
        queryParams.push(`name=${query}`)
    }

    if (aliveOnly) {
        queryParams.push('status=alive')
    }

    const searchQuery = queryParams.length > 0 ? `?${queryParams.join('&')}` : ''

    axios.get(`https://rickandmortyapi.com/api/character/${searchQuery}`)
        .then(response => {
            dispatch(actionCreators.charactersListLoaded(response.data.results))
        })
        .catch((err) => {
            dispatch(actionCreators.charactersListLoadFailed())
        })
}
