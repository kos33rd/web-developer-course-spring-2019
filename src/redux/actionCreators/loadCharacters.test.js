import axios from 'axios';
import moxios from 'moxios';
import { loadCharactersActionCreator } from "./loadCharacters"

// jest.mock('axios')
//
// const resp = { data: { results: [] } }
// axios.get = jest.fn(() => {
//     console.log('Get called')
//     return {
//         then: jest.fn((fn) => {
//             fn(resp)
//             return {
//                 catch: jest.fn()
//             }
//         })
//     }
// })

const dispatch = jest.fn()


describe('action loadCharactersActionCreator', () => {

    beforeEach(() => {
        moxios.install()

    })

    it('tests moxios', (done) => {

        let onFulfilled = jest.fn()

        moxios.stubRequest('/say/hello', {
            status: 200,
            responseText: 'hello'
        })

        axios.get('/say/hello').then(onFulfilled)

        moxios.wait(function () {
            expect(onFulfilled).toBeCalled()
            console.log('Kekeke')
            done()
        })
    })

    it('works fine', (done) => {

        moxios.stubRequest('https://rickandmortyapi.com/api/character/', {
            status: 200,
            responseText: {
                results: [{ test: 'character' }]
            }
        })

        const dispatcher = loadCharactersActionCreator()
        expect(typeof dispatcher).toBe('function')

        dispatcher(dispatch)

        moxios.wait(function () {
            const req = moxios.requests.mostRecent()
            expect(dispatch).toBeCalledTimes(1)
            expect(dispatch.mock.calls[0][0]).toMatchSnapshot()
            done()
        })
    })

    afterEach(() => {
        moxios.uninstall()
    })

})
