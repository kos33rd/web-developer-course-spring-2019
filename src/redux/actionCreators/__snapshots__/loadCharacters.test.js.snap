// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action loadCharactersActionCreator works fine 1`] = `
Object {
  "characters": Array [
    Object {
      "test": "character",
    },
  ],
  "type": "CHARACTERS_LIST_LOADED",
}
`;
